src/js/links.js

```diff
@@ -0,0 +1,251 @@
+      }
+      sites.add(site);
+
+      return !gBlockedLinks.isBlocked(link) && !gPinnedLinks.isPinned(link);
+    });
+
+    // Try to fill the gaps between pinned links.
+    for (var i = 0; i < pinnedLinks.length && links.length; i++) {
+      if (!pinnedLinks[i]) {
+        pinnedLinks[i] = links.shift();
+      }
+    }
+
+    // Append the remaining links if any.
+    if (links.length) {
+      pinnedLinks = pinnedLinks.concat(links);
+    }
+
+    for (link of pinnedLinks) {
+      if (link) {
+        link.baseDomain = ProviderManager.extractSite(link.url);
+      }
+    }
+
+    // Get the set of enhanced links (if any) from the Directory Links Provider.
+    var enhancedLinks = [];
+    for (link of pinnedLinks) {
+      if (link) {
+        enhancedLinks.push(DirectoryLinksProvider.getEnhancedLink(link));
+      }
+    }
+
+    return {links: pinnedLinks, enhancedLinks};
+  },
+
+  _incrementSiteMap: function(map, link) {
+    if (gBlockedLinks.isBlocked(link)) {
+      // Don't count blocked URLs.
+      return;
+    }
+    var site = ProviderManager.extractSite(link.url);
+    map.set(site, (map.get(site) || 0) + 1);
+  },
+
+  /**
+   * Calls getLinks on the given provider and populates our cache for it.
+   *
+   * @param {Provider} aProvider The provider whose cache will be populated.
+   * @param {Object} cache The provider cache to be populated.
+   * @param {Boolean} aForce When true, populates the provider's cache even when it's
+   *               already filled.
+   */
+  _populateProviderCache: async(function*(aProvider, cache, aForce) {
+    var createCache = !cache;
+    if (createCache) {
+      cache = {
+        // Start with a resolved promise.
+        populatePromise: new Promise(resolve => resolve()),
+      };
+      Links._providers.set(aProvider, cache);
+    }
+    // Chain the populatePromise so that calls are effectively queued.
+    cache.populatePromise = cache.populatePromise.then(() => {
+      return new Promise(resolve => {
+        if (!createCache && !aForce) {
+          resolve();
+          return;
+        }
+        aProvider.getLinks().then((links) => {
+          // Filter out null and undefined links so we don't have to deal with
+          // them in getLinks when merging links from providers.
+          links = links.filter((link) => !!link);
+          cache.sortedLinks = links;
+          cache.siteMap = links.reduce((map, link) => {
+            Links._incrementSiteMap(map, link);
+            return map;
+          }, new Map());
+          cache.linkMap = links.reduce((map, link) => {
+            map.set(link.url, link);
+            return map;
+          }, new Map());
+          resolve();
+        });
+      });
+    });
+    yield cache.populatePromise;
+  }),
+
+  /**
+   * Populates the cache with fresh links from the providers.
+   *
+   * @param {Boolean} aForce When true, populates the cache even when it's already filled.
+   */
+  populateCache: async(function*(aForce) {
```

Comment:
With the changes above, then `populateCache()` might not have to be an async function. If so, all that would be needed would be:
```JS
Array.from(this._providers.keys())
  .forEach(provider => this._populateProviderCache(provider, options)) 
```

Or maybe it could all the swallowed up by _populateProviderCache() and that interface could become public?